File: php/src/HorizontalAlign.php
<?php

namespace Vestaboard\Vbml;

/**
 * Port of horizontalAlign.ts
 */
class HorizontalAlign
{
    /**
     * @param int      $width   The width of each row
     * @param string   $justify 'left'|'right'|'justified'|'center' (default)
     * @param int[][]  $codes   Array of rows, each row an array of ints
     * @return int[][]
     */
    public static function align(int $width, string $justify, array $codes): array
    {
        switch ($justify) {
            case 'left':
                return array_map(function (array $row) {
                    return self::removeExtraSpace($row)['row'];
                }, $codes);

            case 'right':
                return array_map(function (array $row) use ($width) {
                    $reversed = array_reverse($row);
                    $trimmed = self::removeExtraSpace($reversed)['row'];
                    $result = array_fill(0, $width, CharacterCode::Blank);
                    for ($i = 0; $i < $width; $i++) {
                        $srcIndex = $width - 1 - $i;
                        $result[$i] = $trimmed[$srcIndex] ?? CharacterCode::Blank;
                    }
                    return $result;
                }, $codes);

            case 'justified':
                $rows = array_map(fn($row) => self::removeExtraSpace($row)['row'], $codes);
                $longestRow = 0;
                foreach ($rows as $row) {
                    if (count($row) > $longestRow) {
                        $longestRow = count($row);
                    }
                }
                $longestRow = $longestRow - 1;
                $paddingRight = (int)floor(($width - $longestRow) / 2);
                $paddingLeft = $width - ($longestRow + ($paddingRight + 1));
                $padding = $paddingRight > $paddingLeft ? $paddingLeft : $paddingRight;

                return array_map(function (array $row) use ($width, $padding) {
                    $result = array_fill(0, $width, CharacterCode::Blank);
                    for ($i = 0; $i < $width; $i++) {
                        $srcIndex = $i - $padding;
                        $result[$i] = $row[$srcIndex] ?? CharacterCode::Blank;
                    }
                    return $result;
                }, $rows);

            default: // center
                return array_map(function (array $row) use ($width) {
                    $reversed = array_reverse($row);
                    $trimmedReversed = self::removeExtraSpace($reversed)['row'];
                    $trimmed = array_reverse($trimmedReversed);
                    $paddingLeft = (int)floor(($width - count($trimmed)) / 2);
                    $result = array_fill(0, $width, CharacterCode::Blank);
                    for ($i = 0; $i < $width; $i++) {
                        $srcIndex = $i - $paddingLeft;
                        $result[$i] = $trimmed[$srcIndex] ?? CharacterCode::Blank;
                    }
                    return $result;
                }, $codes);
        }
    }

    private static function removeExtraSpace(array $row): array
    {
        $resultRow = [];
        $extraSpace = true;
        foreach ($row as $cur) {
            if ($cur === CharacterCode::Blank && $extraSpace) {
                // skip leading blank
            } else {
                $resultRow[] = $cur;
                $extraSpace = false;
            }
        }
        return ['row' => $resultRow, 'extraSpace' => $extraSpace];
    }
}

File: python/pyvbml/horizontal_align.py
"""Horizontal align.

Port of Vestaboard/vbml/src/horizontalAlign.ts
"""

from __future__ import annotations

import math

from .character_codes import CharacterCode
from .types import Justify


def _remove_extra_space(row: list[int]) -> list[int]:
    """Strip leading blank characters (CharacterCode.Blank == 0)."""
    result: list[int] = []
    extra_space = True
    for code in row:
        if code == CharacterCode.BLANK and extra_space:
            continue
        result.append(code)
        extra_space = False
    return result


def horizontal_align(
    width: int, justify: Justify, codes: list[list[int]]
) -> list[list[int]]:
    """Horizontal align."""
    if justify == Justify.LEFT:
        return [_remove_extra_space(row) for row in codes]

    if justify == Justify.RIGHT:
        out = []
        for row in codes:
            stripped = list(reversed(_remove_extra_space(list(reversed(row)))))
            result: list[int] = [int(CharacterCode.BLANK)] * width
            for i, val in enumerate(reversed(stripped)):
                if i < width:
                    result[width - 1 - i] = val
            out.append(result)
        return out

    if justify == Justify.JUSTIFIED:
        rows = [_remove_extra_space(row) for row in codes]
        longest = max((len(r) for r in rows), default=0) - 1
        padding_right = math.floor((width - longest) / 2)
        padding_left = width - (longest + (padding_right + 1))
        padding = min(padding_right, padding_left)
        return [
            [
                row[i - padding]
                if 0 <= i - padding < len(row)
                else int(CharacterCode.BLANK)
                for i in range(width)
            ]
            for row in rows
        ]

    # default / center
    out = []
    for row in codes:
        stripped = list(reversed(_remove_extra_space(list(reversed(row)))))
        padding_left = math.floor((width - len(stripped)) / 2)
        out.append(
            [
                stripped[i - padding_left]
                if 0 <= i - padding_left < len(stripped)
                else int(CharacterCode.BLANK)
                for i in range(width)
            ]
        )
    return out
